### DISCUSSION TITLE
BLOB, TEXT, GEOMETRY or JSON column 'emails' can't have a default value

### DISCUSSION THREAD

**Discussion by Overload119** [planetscale/discussion#101] [Bug Reports] [answered]
I thought that it would be supported as MySQL supports it.

**Comment by jgreet** [comment]
We'll take a look at this, could be an incompatibility with Vitess.

**Comment by Overload119** [comment, marked as answer]
FTR - I fixed this by changing this line in my migration file.

```
t.json "emails", default: -> { "(json_array())" }, null: false
```

**Reply by jgreet** [reply]
Glad you have a workaround. 

There's definitely a bug in the query parser that can strip the parenthesis from a default value. I'll update this thread when we have a fix.

**Reply by jgreet** [reply]
Issue and potential fix in Vitess: https://github.com/vitessio/vitess/issues/8982

**Comment by ImLunaHey** [comment]
@jgreet there seems to be some disconnect somewhere.

My migrations ran and created a table with a json field that has a default.
In the tables area I see this for the table.

```sql
CREATE TABLE `audit_logs` (
	`id` varchar(36) NOT NULL DEFAULT (uuid()),
	`guild_id` varchar(36) NOT NULL,
	`enabled` tinyint(1) DEFAULT '1',
	`channel_id` varchar(36) NOT NULL,
	`ignored_actions` json DEFAULT (_utf8mb4 '[]'),
	`ignore_bots` tinyint(1) DEFAULT '0',
	`ignored_roles` json DEFAULT (_utf8mb4 '[]'),
	`ignored_users` json DEFAULT (_utf8mb4 '[]'),
	`ignored_channels` json DEFAULT (_utf8mb4 '[]'),
	PRIMARY KEY (`id`),
	UNIQUE KEY `guild_id_channel_id_unique` (`guild_id`, `channel_id`)
) ENGINE InnoDB,
  CHARSET utf8mb4,
  COLLATE utf8mb4_0900_ai_ci;
```

But if I try and alter the table it won't let me.
![image](https://user-images.githubusercontent.com/105922175/235330113-7aa69f7c-43f7-40af-9a8e-3309e7e509d8.png)

**Reply by naishe** [reply]
@jgreet -- I am facing the same issue.

```text
> prisma db push

Prisma schema loaded from prisma/schema.prisma
Datasource "db": MySQL database "myenv" at "aws.connect.psdb.cloud"
Error: target: myenv.-.primary: vttablet: rpc error: code = InvalidArgument desc = BLOB, TEXT, GEOMETRY or JSON column 'marketplaceCountryIds' can't have a default value (errno 1101) (sqlstate 42000) (CallerID: ****): Sql: "alter table AdConfig modify column marketplaceCountryIds TEXT not null default '[\\\"IN\\\"]'", BindVars: {REDACTED}
   0: sql_schema_connector::apply_migration::migration_step
           with step=AlterTable(AlterTable { table_ids: MigrationPair { previous: TableId(4), next: TableId(6) }, changes: [AlterColumn(AlterColumn { column_id: MigrationPair { previous: TableColumnId(56), next: TableColumnId(63) }, changes: ColumnChanges { type_change: None, changes: BitFlags<ColumnChange>(0b10, Default) }, type_change: None })] })
             at schema-engine/connectors/sql-schema-connector/src/apply_migration.rs:21
   1: sql_schema_connector::apply_migration::apply_migration
             at schema-engine/connectors/sql-schema-connector/src/apply_migration.rs:10
   2: schema_core::state::SchemaPush
             at schema-engine/core/src/state.rs:433
```

**Reply by jgreet** [reply]
This is the result of a recent [fix in Vitess](https://github.com/vitessio/vitess/pull/13730). Previously, Vitess would accept a create or alter statement for a column like `test text not null default ""` and rewrite the default value as an expression `test text not null default ("")`.


The current defaults on your TEXT columns are expressions, `(_utf8mb4'[\"IN\"]')` for example, not string "values" like `'["IN"]'`.
 
The [MySQL 8 documentation states](https://dev.mysql.com/doc/refman/8.0/en/blob.html) that "BLOB and TEXT columns cannot have DEFAULT values" so the new behavior is more compatible with MySQL and unlikely to be reverted.
 
To proceed without changing the schema in your database, you could make the column defaults specified in `schema.prisma` match the current expressions. Using the example in your error message...
 
Instead of something like:

```js
marketplaceCountryIds String @default("[\"IN\"]") @db.Text
```
 
Use:
 
```ls
marketplaceCountryIds String @default(dbgenerated("(_utf8mb4'[\"IN\"]')")) @db.Text
```
 
I would recommend running `prisma db pull --print` to see the exact definitions for your current columns.

**Reply by janpio** [reply]
(And please let us at Prisma know if `db pull` does not return a valid, working value here. Thanks.)

**Reply by naishe** [reply]
Thanks for bringing me up to date with the latest changes. 

@janpio, `db pull` returns exactly what @jgreet has mentioned and editing my `schema.prisma` file to the new format fixes the error. Here's the diff

```diff
-  marketplaceCountryIds String                @default("[\"IN\"]") @db.Text
+  marketplaceCountryIds String                @default(dbgenerated("(_utf8mb4\\'[\"IN\"]\\')")) @db.Text
```

and for empty string

```diff
-  sessionKey  String    @default("") @db.Text
+  sessionKey  String    @default(dbgenerated("(_utf8mb4\\'\\')")) @db.Text
```

This is needed only in the fields that have `@db.Text` annotations.
